javascript/ql/src/semmle/javascript/GlobalAccessPaths.qll

```diff
@@ -96,90 +160,205 @@ module GlobalAccessPath {
    * for nodes that transitively flow there.
    *
-   * For example, the class nodes below all map to `foo.bar`:
+   * For example, the class nodes below all map to `foo.bar` relative to `x`:
    * ```
-   * foo.bar = class {};
-   *
-   * foo = { bar: class {} };
-   *
-   * (function(f) {
-   *   f.bar = class {}
-   *  })(foo = foo || {});
+   * function f(x) {
+   *   x.foo.bar = class {};
+   *   x.foo = { bar: class() };
+   *   let alias = x;
+   *   alias.foo.bar = class {};
+   * }
    * ```
    */
   cached
-  string fromRhs(DataFlow::Node node) {
-    exists(DataFlow::SourceNode base, string baseName, string name |
-      node = base.getAPropertyWrite(name).getRhs() and
-      result = baseName + "." + name
+  private string fromRhs(DataFlow::Node node, Root root) {
+    exists(DataFlow::PropWrite write, string baseName |
+      node = write.getRhs() and
+      result = join(baseName, write.getPropertyName())
     |
-      baseName = fromReference(base)
+      baseName = fromReference(write.getBase(), root)
       or
-      baseName = fromRhs(base)
+      baseName = fromRhs(write.getBase(), root)
     )
     or
     exists(GlobalVariable var |
       node = var.getAnAssignedExpr().flow() and
-      result = var.getName()
+      result = var.getName() and
+      root.isGlobal()
     )
     or
     exists(FunctionDeclStmt fun |
       node = DataFlow::valueNode(fun) and
-      result = fun.getId().(GlobalVarDecl).getName()
+      result = fun.getId().(GlobalVarDecl).getName() and
+      root.isGlobal()
     )
     or
     exists(ClassDeclStmt cls |
       node = DataFlow::valueNode(cls) and
-      result = cls.getIdentifier().(GlobalVarDecl).getName()
+      result = cls.getIdentifier().(GlobalVarDecl).getName() and
+      root.isGlobal()
     )
     or
     exists(EnumDeclaration decl |
       node = DataFlow::valueNode(decl) and
-      result = decl.getIdentifier().(GlobalVarDecl).getName()
+      result = decl.getIdentifier().(GlobalVarDecl).getName() and
+      root.isGlobal()
     )
     or
     exists(NamespaceDeclaration decl |
       node = DataFlow::valueNode(decl) and
-      result = decl.getId().(GlobalVarDecl).getName()
+      result = decl.getId().(GlobalVarDecl).getName() and
+      root.isGlobal()
     )
   }
 
   /**
-   * Gets the global access path referenced by or assigned to `node`.
+   * Gets a node that refers to the given access path relative to the given `root` node,
+   * or `root` itself if the access path is empty.
+   *
+   * This works for direct references as well as for aliases established through local data flow.
+   *
+   * For example:
+   * ```
+   * function f(x) {
+   *   let a = x.f.g; // reference to (x, "f.g")
+   *   let b = a.h;   // reference to (x, "f.g.h")
+   * }
+   * ```
    */
-  string getAccessPath(DataFlow::Node node) {
-    result = fromReference(node)
+  DataFlow::Node getAReferenceTo(Root root, string path) {
+    path = fromReference(result, root) and
+    not root.isGlobal()
```

Comment:
Is `not root.isGlobal()` used to help the optimizer? Or is it just done to encourage not using `DataFlow::globalAccessPathRootPseudoNode()` explicitly?

Reply:
It's mainly to avoid exposing the pseudo-node, yes.